File: app/src/main/java/com/kickstarter/ui/activities/ProjectUpdatesActivity.java
package com.kickstarter.ui.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.kickstarter.R;
import com.kickstarter.libs.BaseActivity;
import com.kickstarter.libs.RecyclerViewPaginator;
import com.kickstarter.libs.SwipeRefresher;
import com.kickstarter.libs.qualifiers.RequiresActivityViewModel;
import com.kickstarter.models.Project;
import com.kickstarter.models.Update;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.adapters.UpdatesAdapter;
import com.kickstarter.ui.toolbars.KSToolbar;
import com.kickstarter.viewmodels.ProjectUpdatesViewModel;

import org.jetbrains.annotations.NotNull;

import butterknife.Bind;
import butterknife.BindString;
import butterknife.ButterKnife;
import rx.android.schedulers.AndroidSchedulers;

import static com.kickstarter.libs.rx.transformers.Transformers.observeForUI;
import static com.kickstarter.libs.utils.TransitionUtils.slideInFromLeft;

@RequiresActivityViewModel(ProjectUpdatesViewModel.ViewModel.class)
public class ProjectUpdatesActivity extends BaseActivity<ProjectUpdatesViewModel.ViewModel> implements UpdatesAdapter.Delegate {
  private UpdatesAdapter adapter;
  private RecyclerViewPaginator recyclerViewPaginator;
  private SwipeRefresher swipeRefresher;

  protected @Bind(R.id.updates_toolbar) KSToolbar updatesToolbar;
  protected @Bind(R.id.updates_swipe_refresh_layout) SwipeRefreshLayout swipeRefreshLayout;
  protected @Bind(R.id.updates_recycler_view) RecyclerView recyclerView;

  protected @BindString(R.string.project_subpages_menu_buttons_updates) String updatesTitleString;

  @Override
  protected void onCreate(final @Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_updates);
    ButterKnife.bind(this);

    this.adapter = new UpdatesAdapter(this);
    this.recyclerView.setAdapter(this.adapter);
    this.recyclerView.setLayoutManager(new LinearLayoutManager(this));

    this.recyclerViewPaginator = new RecyclerViewPaginator(this.recyclerView, this.viewModel.inputs::nextPage, this.viewModel.outputs.isFetchingUpdates());
    this.swipeRefresher = new SwipeRefresher(
            this, this.swipeRefreshLayout, this.viewModel.inputs::refresh, this.viewModel.outputs::isFetchingUpdates
    );
    this.updatesToolbar.setTitle(this.updatesTitleString);

    this.viewModel.outputs.startUpdateActivity()
      .compose(bindToLifecycle())
      .compose(observeForUI())
      .subscribe(pu -> this.startUpdateActivity(pu.first, pu.second));

    this.viewModel.outputs.projectAndUpdates()
      .compose(bindToLifecycle())
      .observeOn(AndroidSchedulers.mainThread())
      .subscribe(this.adapter::takeData);
  }

  @Override
  protected void onDestroy() {
    super.onDestroy();

    this.recyclerViewPaginator.stop();
    this.recyclerView.setAdapter(null);
  }

  @Override
  public void updateClicked(final @NotNull Update update) {
    this.viewModel.inputs.goToUpdate(update);
  }

  private void startUpdateActivity(final @NonNull Project project, final @NonNull Update update) {
    final Intent intent = new Intent(this, UpdateActivity.class)
      .putExtra(IntentKey.PROJECT, project)
      .putExtra(IntentKey.UPDATE, update);
    startActivityWithTransition(intent, R.anim.slide_in_right, R.anim.fade_out_slide_out_left);
  }

  @Override
  protected @Nullable Pair<Integer, Integer> exitTransition() {
    return slideInFromLeft();
  }

}

File: app/src/main/java/com/kickstarter/ui/activities/ProjectUpdatesActivity.kt
package com.kickstarter.ui.activities

import android.content.Intent
import android.os.Bundle
import androidx.activity.addCallback
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isGone
import androidx.recyclerview.widget.LinearLayoutManager
import com.kickstarter.R
import com.kickstarter.databinding.ActivityUpdatesBinding
import com.kickstarter.libs.recyclerviewpagination.RecyclerViewPaginatorV2
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.getEnvironment
import com.kickstarter.models.Project
import com.kickstarter.models.Update
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.adapters.UpdatesAdapter
import com.kickstarter.ui.extensions.finishWithAnimation
import com.kickstarter.ui.extensions.setUpConnectivityStatusCheck
import com.kickstarter.ui.extensions.startActivityWithTransition
import com.kickstarter.utils.WindowInsetsUtil
import com.kickstarter.viewmodels.ProjectUpdatesViewModel.Factory
import com.kickstarter.viewmodels.ProjectUpdatesViewModel.ProjectUpdatesViewModel
import io.reactivex.disposables.CompositeDisposable

class ProjectUpdatesActivity : AppCompatActivity(), UpdatesAdapter.Delegate {
    private val adapter: UpdatesAdapter = UpdatesAdapter(this)
    private lateinit var recyclerViewPaginator: RecyclerViewPaginatorV2

    private lateinit var binding: ActivityUpdatesBinding

    private lateinit var viewModelFactory: Factory
    private val viewModel: ProjectUpdatesViewModel by viewModels {
        viewModelFactory
    }

    private val disposables = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityUpdatesBinding.inflate(layoutInflater)
        WindowInsetsUtil.manageEdgeToEdge(
            window,
            binding.root
        )
        setContentView(binding.root)

        setUpConnectivityStatusCheck(lifecycle)

        this.getEnvironment()?.let { env ->
            viewModelFactory = Factory(env, intent)
        }

        binding.updatesRecyclerView.adapter = adapter
        binding.updatesRecyclerView.layoutManager = LinearLayoutManager(this)

        recyclerViewPaginator = RecyclerViewPaginatorV2(
            binding.updatesRecyclerView,
            { viewModel.inputs.nextPage() },
            viewModel.outputs.isFetchingUpdates()
        )

        binding.updatesSwipeRefreshLayout.setOnRefreshListener {
            viewModel.inputs.refresh()
        }

        viewModel.outputs.isFetchingUpdates()
            .compose(Transformers.observeForUIV2())
            .subscribe {
                // - Hides loading spinner from SwipeRefreshLayout according to isFetchingUpdates
                binding.updatesSwipeRefreshLayout.isRefreshing = it
            }
            .addToDisposable(disposables)

        binding.updatesToolbar.commentsToolbar.setTitle(getString(R.string.project_subpages_menu_buttons_updates))

        viewModel.outputs.horizontalProgressBarIsGone()
            .compose(Transformers.observeForUIV2())
            .subscribe {
                binding.updatesProgressBar.isGone = it
            }
            .addToDisposable(disposables)

        viewModel.outputs.startUpdateActivity()
            .compose(Transformers.observeForUIV2())
            .subscribe { startUpdateActivity(it.first, it.second) }
            .addToDisposable(disposables)

        viewModel.outputs.projectAndUpdates()
            .compose(Transformers.observeForUIV2())
            .subscribe { adapter.takeData(it) }
            .addToDisposable(disposables)

        this.onBackPressedDispatcher.addCallback {
            finishWithAnimation()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        recyclerViewPaginator.stop()
        binding.updatesRecyclerView.adapter = null
        disposables.clear()
    }

    override fun updateCardClicked(update: Update) {
        viewModel.inputs.updateClicked(update)
    }

    private fun startUpdateActivity(project: Project, update: Update) {
        val intent = Intent(this, UpdateActivity::class.java)
            .putExtra(IntentKey.PROJECT, project)
            .putExtra(IntentKey.UPDATE, update)
        startActivityWithTransition(intent, R.anim.slide_in_right, R.anim.fade_out_slide_out_left)
    }
}
